src/coreclr/System.Private.CoreLib/src/System/Threading/GreenThreadTasks.CoreCLR.cs

```diff
@@ -0,0 +1,111 @@
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "GreenThread_ResumeThread")]
+        private static unsafe partial SuspendedThread* GreenThread_ResumeThread(SuspendedThread* suspendedThread);
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "GreenThread_Yield")]
+        [return: MarshalAs(UnmanagedType.I1)]
+        private static unsafe partial bool GreenThread_Yield();
+
+        [UnmanagedCallersOnly]
+        private static unsafe void GreenThreadStartFunc(void* argument)
+        {
+            Action action = Unsafe.AsRef<Action>(argument);
+            action();
+        }
+
+        private static class GreenThreadStatics
+        {
+            [ThreadStatic]
+            public static Task? t_TaskToWaitFor;
+        }
+
+        internal static unsafe void GreenThreadExecutorFunc(object? obj)
+        {
+            Thread.t_IsGreenThread = true;
+            try
+            {
+                var action = (Action?)obj;
+                var suspendedThread = GreenThread_StartThread(&GreenThreadStartFunc, Unsafe.AsPointer(ref action));
+                Debug.Assert((GreenThreadStatics.t_TaskToWaitFor != null) == (suspendedThread != null));
+                if (suspendedThread != null)
+                {
+                    Debug.Assert(GreenThreadStatics.t_TaskToWaitFor != null);
+                    var suspendedThreadObj = new ResumeSuspendedThread(suspendedThread);
+                    GreenThreadStatics.t_TaskToWaitFor.GetAwaiter().OnCompleted(suspendedThreadObj.Resume);
+                }
+            }
+            finally
+            {
+                GreenThreadStatics.t_TaskToWaitFor = null;
+                Thread.t_IsGreenThread = false;
+            }
+        }
+
+        private sealed unsafe class ResumeSuspendedThread
+        {
+            private SuspendedThread* _suspendedThread;
+            public ResumeSuspendedThread(SuspendedThread* suspendedThread)
+            {
+                _suspendedThread = suspendedThread;
+            }
+
+            public void Resume()
+            {
+                Thread.t_IsGreenThread = true;
+                try
+                {
+                    _suspendedThread = GreenThread_ResumeThread(_suspendedThread);
+                    Debug.Assert((GreenThreadStatics.t_TaskToWaitFor != null) == (_suspendedThread != null));
+                    if (_suspendedThread != null)
+                    {
+                        Debug.Assert(GreenThreadStatics.t_TaskToWaitFor != null);
+                        GreenThreadStatics.t_TaskToWaitFor.GetAwaiter().OnCompleted(Resume);
+                    }
+                }
+                finally
+                {
+                    GreenThreadStatics.t_TaskToWaitFor = null;
+                    Thread.t_IsGreenThread = false;
+                }
+            }
+        }
+
+        static partial void RunOnActualGreenThread(Action action, ref bool ranAsActualGreenThread)
```

Comment:
Is there a reason to use `ref bool` instead of `out bool`?

Reply:
Yes, partial methods are not permitted to have out parameters. Effectively, if the method implementation does not exist, then the method isn't called and any out parameters would have no meaning.